src/gamestates/gameState.lua

```diff
@@ -1,43 +1,15 @@
-local GameState = {}
-GameState.__index = GameState
+local enabledFunctions = require("gamestates/enabledFunctions")
 
-function GameState.load()
-end
-
-function GameState.update()
-end
-
-function GameState.draw()
-end
-
-function GameState.keypressed()
-end
-
-function GameState.keyreleased()
-end
-
-function GameState.mousepressed()
-end
-
-function GameState.mousereleased()
-end
-
-function GameState.joystickpressed()
-end
+local function emptyFunction() end
 
-function GameState.joystickreleased()
+local function index(t, key)
+	return getmetatable(t)[key] or (enabledFunctions[key] and emptyFunction)
```

Comment:
Shouldn't this be `getmetatable(t)__index[key]` ?

Reply:
Wait, that would be an infinite loop. Where is the reference to GameState?

Reply:
Or...shouldn't it be `t[key]`?

Reply:
I guess I'm not sure what the part before the or is for.

Reply:
It works the way I intended/planned if you don't understand why/how it works I can explain it.

Reply:
Can you explain it?

Reply:
Relevant code: 
```lua

local function index(t, key)
	return getmetatable(t)[key] or (enabledFunctions[key] and emptyFunction)
end

local GameState = setmetatable({__index = index}, {__call = call})

return GameState
```
When "{currentState}.{functionname}()" is called and that function is not defined by the current state, then the index function is called since it is set as the __index of GameState. Once it is called the current state is handed is as a table "t" and the name of the function as "key". It then gets the metatable of "t" which is always "GameState" and checks to see if GameState has the function defined(Currently non are defined). If it is not defined in gamestate and it it is an enabled function then it returns a empty function to prevent an error.

Reply:
Compare it to the function in class.lua

Reply:
I just tested this and it calls "GameState.draw" if you comment out "InGame.draw".
```lua
local GameState = setmetatable({__index = index}, {__call = call})

function GameState.draw()
	print("GameState.draw")
end

return GameState
```

Reply:
Ok, so the answer is that the metatable of t is GameState, and the part before the `or` the function in GameState if it exists.

Reply:
I think that part (`getmetatable(t)[key] or `) is redundant because Lua doesn't use `__index` if the key already exists.

Reply:
Oh, I see. You're not setting GameState as the __index of the gamestates.

So I commented out `inGame.draw()` and defined `GameState.draw()` as you mentioned in your example. Here's what happens, as I understand it:
1. Lua checks to see if there's a draw function in inGame. There isn't.
2. It looks up the __index of inGame's metatable. inGame's metatable is GameState, which doesn't have an __index method.
3. Since GameState doesn't have an __index method, Lua uses the __index of GameState to look it up. It's the index function that you defined.
4. The arguments to `index` are: `index(inGame, "draw")`.
5. `getmetatable(t)[key]` == `getmetatable(inGame)["draw"]` == `GameState["draw"]`, which is a function that is not nil, so it returns this function without evaluating the rest of the line.